<img src="https://ei.marketwatch.com/Multimedia/2016/08/01/Photos/NS/MW-ES907_didi_u_20160801154402_NS.jpg?uuid=4c17b2d4-5820-11e6-8568-0015c588dfa6">

<h1>Ride-Hailing Platforms</h1>
<br>
<font size="+1">
    <ul>
        <li>Ride-hailing platforms such as Uber, Lyft, and DiDi have achieved explosive growth and reshaped urban transportation.</li>
        <br>
        <ul>
            <li>For example, Uber has completed over 10 billion trips globally and is active in over 80 countries and 700 cities since its initial launch in 2009.</li>
            <br>
            <li>By some estimates, the global ride-hailing industry is projected to grow to a 285-billion dollar total market value by 2030.</li>
            <br>
        </ul>
        <li>Ride-hailing platforms connect riders and drivers via a <i>centralized</i> and <i>automated matching and pricing system</i> in a <b>two-sided marketplace</b>.</li>
        <br>
        <li>Ride-hailing platforms have three components:</li>
        <br>
        <ul>
            <li>driver app (for drivers to offer services and communicate with their customers),</li>
            <br>
            <li>rider app (for customers to book and track their journeys and select vehicle types),</li>
            <br>
            <li>and a dispatch system (a system that connects the driver and the customer via their mobile phones).</li>
            <br>
        </ul>
        <li>Relative to taxi services, drivers participating in ride-hailing platforms spend a higher proportion of their time delivering rides.</li>
        <br>
        <ul>
            <li>This is known as having a higher <b><i>capacity utilization rate</i></b>.</li>
            <br>
            <li>In other words, UberX drivers spend less time waiting between ride requests or driving to pick up riders, compared to taxi drivers.</li>
            <br>
            <li>This is typically true on the driver side as well.</li>
            <br>
            <li>Many studies have been conducted and they have found that the average rider waiting time between request and pickup (known as <b><i>rider waiting time</i></b>) can be much lower in a ride-hailing system compared to a street-hailing system.</li>
            <br>
        </ul>
        <li style="color:blue">The are two key levers in building a successful ride-hailing platform:</li>
        <br>
        <ul style="color:blue">
            <li>matching algorithms, which help to efficiently match riders and drivers,</li>
            <br>
            <li>and dynamic pricing algorithms, which help to keep the supply of drivers and demand of riders in balance.</li>
            <br>
        </ul>
        <li>Matching refers to the process of dispatching available drivers to pick up riders.</li>
        <br>
        <ul>
            <li><img src="https://www.uber-assets.com/image/upload/f_auto,q_auto:eco,c_fill,w_956,h_637/v1611193266/assets/58/69b90b-b3a6-412f-a45e-4a28026cd86a/original/2_Better-matching-matters-grey.jpg" width="300"></li>
            <br>
        </ul>
        <li>Dynamic pricing refers to the adjustment of prices for rides over time based on real-time demand and supply conditions.</li>
        <br>
        <ul>
            <li>Dynamic pricing is called <i>surge pricing</i> by Uber and <i>prime time</i> by Lyft.</li>
            <br>
            <li><img src="https://pbs.twimg.com/media/FQJrV5LXIAAclOb?format=jpg&name=900x900" height="300" width="300"></li>
            <br>
        </ul>
        <li>A successful implementation of matching and pricing can be leveraged to create an experience with <b>low waiting times</b> for both riders and drivers.</li>
        <br>
        <li style="color:dodgerblue">As an aside:</li>
        <br>
        <ul style="color:dodgerblue">
            <li>Matching and dynamic pricing algorithms have been leveraged by many other online platforms that connect service providers and consumers.</li>
            <br>
            <li>For example,</li>
            <br>
            <ul>
                <li><a href="https://www.taskrabbit.com/blog/unveiling-the-new-taskrabbit">TaskRabbit</a> for chores and home projects,</li>
                <br>
                <li><a href="https://www.ebay.com/help/selling/listings/listing-tips/optimising-listings-best-match?id=4166">eBay</a> for e-commerce,</li>
                <br>
                <li><a href="https://www2.aueb.gr/users/mitsopoulou/files/Parking_Percom_camera_ready(1).pdf">SpotHero</a> for parking spaces,</li>
                <br>
                <li>StubHub for concert and sports game tickets,</li>
                <br>
                <li>Turo and Getaround for car-sharing,</li>
                <br>
                <li>and <a href="https://blogs.cornell.edu/info4220/2016/03/18/how-airbnb-uses-matching-markets-to-ensure-host-preferences/">Airbnb</a> for hospitality services.</li>
                <br>
            </ul>
            <li>In contrast to ride-hailing, these services provide matching <i>suggestions</i> and price <i>guidance</i> via recommendation systems, but the final decisions are made by the participants on the platform as opposed to the platform itself.</li>
            <br>
            <li>Food and grocery delivery platforms (such as UberEats, Instacard, or Grubhub) are more similar to ride-hailing platforms, but they have more time and flexibility for dispatching due to a lower level of urgency for the service.</li>
            <br>
            <li>The pricing and matching decisions for ride-hailing services need to be resolved in real-time as riders and drivers are sensitive in time.</li>
            <br>
            <li>Matching and pricing algorithms are both optimization problems!</li>
            <br>
        </ul>
    </ul>
</font>

<hr style="border:10px solid black">

<h2><a href="https://www.uber.com/blog/engineering-routing-engine/">Brief History of Uber's Routing Engine and Technological Innovation</a></h2>
<br>
<font size="+1">
    <ul>
        <li>When the Uber app was initially launched around 2010, one of the first features of the app was the ETA (estimated time of arrival) between you and the closest driver.</li>
        <br>
        <li>In the early days, Uber used a combination of different routing engines to produce an ETA, including the <a href="http://project-osrm.org/">Open Source Routing Machine (OSRM)</a>, which is an open source optimal routing package and the entire codebase can be found on <a href="https://github.com/Project-OSRM">GitHub</a>.</li>
        <br>
        <ul>
            <li>Also in the early days, Uber didn't have navigation available in the app, so the app was only used to estimate the ETA and to display vehicle locations.</li>
            <br>
            <li>This service of estimating ETAs was called <i>"GoldETA"</i> and it was a model that was built on top of the routing engines and adjusted the original estimates using Uber's historical data of similar routes.</li>
            <br>
            <li>This solution ultimately took into account hundreds of thousands of Uber trips and compared them to the initial routing engine's ETA.</li>
            <br>
            <li>GoldETA worked better than using any single ETA estimate alone.</li>
            <br>
        </ul>
        <li>The main challenge was the <a href="https://en.wikipedia.org/wiki/Cold_start_(recommender_systems)">cold start</a> problem that Uber faced when they launched in new cities.</li>
        <br>
        <ul>
            <li>Specifically, the cold start problem refers to when you want to use a model without having estimated the model on relevant data, that is, your model starts <i>cold</i>.</li>
            <br>
        </ul>
        <li>Additionally, as Uber grew, they needed to add features to their platform that became difficult to implement due to the integration with the open source codebase OSRM.</li>
        <br>
        <li>These challenges led Uber to develop their own routing engine solution.</li>
        <br>
        <li>This required better ETA prediction and better matching and pricing optimization algorithms.</li>
        <br>
        <li>For more on this development, see <a href="https://www.uber.com/blog/engineering-routing-engine/">this link</a>.</li>
        <br>
    </ul>
</font>

<hr style="border:10px solid black">

<h2>Matching in Ride-Hailing</h2>
<br>
<font size="+1">
    <ul>
        <li>Rides can be either non-shared, meaning that the ride is arranged for only one customer group, or shared, meaning that several customer groups with different pickup and dropoff locations share the ride.</li>
        <br>
        <li>In the simpler case of a non-shared ride, drivers participating on the platform cycle through three states sequentially: </li>
        <br>
        <ul>
            <li><i>"open"</i> (waiting to be dispatched),</li>
            <br>
            <li><i>"en route"</i> (on the way to the pickup location),</li>
            <br>
            <li>and <i>"on-trip"</i> (driving riders to their destination).</li>
            <br>
        </ul>
        <li>A request can be matched with a dispatchable driver using very simple algorithms, such as the <b><i>first-dispatch protocol</i></b>.</li>
        <br>
        <ul>
            <li>In the first-dispatch protocol only open drivers are considered as dispatchable.</li>
            <br>
            <li>Each request is immediately assigned to the open driver who is predicted to have the shortest en route time.</li>
            <br>
            <li>This matching algorithm is also called the <i>on-call</i> policy, the <i>closest driver</i> policy, and the <i>committed driver-rider</i> matching protocol.</li>
            <br>
        </ul>
        <li>An alternative matching approach is <i>batching</i> where rider requests are collected for a short time window (typically on the order of seconds), at the end of which an optimization problem is solved to pair each request with an open driver.</li>
        <br>
        <li>If there are riders that are not matched in this batch, then they are carried over and used for the optimization problem in the next batching window.</li>
        <br>
        <li>Batching can reduce the rider waiting time relative to the <i>first-dispatch</i> protocol by consolidating requests and making better use of supply.</li>
        <br>
        <ul>
            <li>Note that the first-dispatch protocol can be viewed as a special case of the batching protocol when each batch consists of exactly one rider request.</li>
            <br>
        </ul>
        <li>Due to its substantial benefits, batching has been implemented widely in major ride-hailing platforms.</li>
        <br>
    </ul>
</font>

<h2>I. Implement the First-Dispatch Routing Protocol with a Maximum Dispatch Radius</h2>
<br>
<font size="+1">
    <ul>
        <li style="color:blue">Recall:</li>
        <br>
        <ul style="color:blue">
            <li>A rider's request can be matched with a dispatchable driver using very simple algorithms, such as the <b><i>first-dispatch protocol</i></b>.</li>
            <br>
            <ul>
                <li>In the first-dispatch protocol only open drivers are considered as dispatchable.</li>
                <br>
                <li><b>Each request is immediately assigned to the open driver who is predicted to have the shortest en route time</b>.</li>
                <br>
                <li>This matching algorithm is also called the <i>on-call</i> policy, the <i>closest driver</i> policy, and the <i>committed driver-rider</i> matching policy.</li>
                <br>
            </ul>
            <li>When this committed driver-rider matching is used by ride-hailing platforms, the dispatched driver could miss the opportunity of being matched with a rider that they drive passed while en route to their matched rider.</li>
            <br>
            <li>To mitigate this inefficiency, some ride-hailing platforms proposed a <i>maximum dispatch radius</i> (MDR)  constraint with carefully selected thresholds.</li>
            <br>
            <li>Maximum dispatch radius means that a dispatch is made only if the en route time is below the threshold.</li>
            <br>
            <li>This prevents the driver from being dispatched to pick up very distant riders.</li>
            <br>
        </ul>
        <li>Assume we want to implement a benchmark matching algorithm solution that is capable of dispatching a single driver, from a collection of available drivers, to a single rider.</li>
        <br>
        <li>Specifically, we are to implement the so-called <i>first-dispatch protocol</i> as an optimization problem given the randomly generated data below.</li>
        <br>
        <li>The data consists of:</li>
        <br>
        <ul>
            <li>A single rider that is requesting a pick-up from a collection of drivers.</li>
            <br>
            <li>A data frame consisting of some measure of physical distance from the rider to each driver.</li>
            <br>
            <li>A maximum dispatch radius (MDR) that represents a cutoff value for drivers to be dispatched to riders.</li>
            <br>
            <ul>
                <li>In other words, if a driver's distance to the rider is larger than the MDR, then the driver cannot be dispatched to the rider's request.</li>
                <br>
            </ul>
            <li>A data frame consisting of the en route time, also known as the expected time of arrival (ETA), which represents the estimated time it will take for a driver to meet a rider once they have been dispatched.</li>
            <br>
        </ul>
        

In [ ]:
# Find a single driver with a minimum ETA
# Find a single driver with minimum ETA that satisfies the MDR (maximum dispatch radius) constraint

In [ ]:
import numpy as np
import pandas as pd

import cvxpy as cp

In [ ]:
rng = np.random.default_rng(42)

In [ ]:
# Number of Riders
num_riders = 1

# Number of drivers
num_drivers = rng.poisson(lam=20)

# Goal is to route DRIVER to RIDER
# Equivalently, to match a rider with a driver

# Physical measure of distance between riders and drivers
distances = pd.DataFrame(rng.uniform(0, 10, (num_riders, num_drivers)),
                         columns=[f'Driver {j}' for j in range(1, num_drivers+1)],
                         index=[f'Rider {i}' for i in range(1,num_riders+1)])

# Map distances to a binary constraint which consists 1 if within MDR and 0 otherwise
# A value of 1 means the driver and rider can be connected.
MDR = 5 # don't dispatch outside of MDR miles
maximum_dispatch_radius = (distances < MDR).apply(lambda x: np.where(x, 1, 0)) # np.where(distances < MDR, 1, 0)

# Temporal measure of distance between riders and drivers
# Expected Time of Arrival
# ETA should be a function of the distance, driver's allowable speed, and some idiosyncratic noise
# (need a more realistic way to assign a variance, maybe as a function of distance)
eta = distances.apply(lambda x: rng.pareto(x))# rng.exponential(x)) 

In [ ]:
eta

Driver 1  Driver 2  Driver 3  Driver 4  Driver 5  Driver 6  Driver 7  \
Rider 1  0.054216  0.011114  0.210143   0.07277  0.052392  0.174482  0.738491   

         Driver 8  Driver 9  Driver 10  ...  Driver 15  Driver 16  Driver 17  \
Rider 1  0.054248  0.132412   0.071678  ...   0.271868   7.841425   0.139952   

         Driver 18  Driver 19  Driver 20  Driver 21  Driver 22  Driver 23  \
Rider 1   0.011628   0.161365   0.465232   0.122537   0.031853   0.042093   

         Driver 24  
Rider 1   0.088881  

[1 rows x 24 columns]

In [ ]:
# distances = distances.transform(lambda x: x+20) # Sanity Check

distances

Driver 1  Driver 2  Driver 3  Driver 4  Driver 5  Driver 6  Driver 7  \
Rider 1  8.585979   6.97368  0.941773  9.756224  7.611397  7.860643  1.281136   

         Driver 8  Driver 9  Driver 10  ...  Driver 15  Driver 16  Driver 17  \
Rider 1  4.503859   3.70798    9.26765  ...   5.545848   0.638173   8.276312   

         Driver 18  Driver 19  Driver 20  Driver 21  Driver 22  Driver 23  \
Rider 1   6.316644   7.580877    3.54526    9.70698   8.931211   7.783835   

         Driver 24  
Rider 1   1.946387  

[1 rows x 24 columns]

In [ ]:
maximum_dispatch_radius

Driver 1  Driver 2  Driver 3  Driver 4  Driver 5  Driver 6  Driver 7  \
Rider 1         0         0         1         0         0         0         1   

         Driver 8  Driver 9  Driver 10  ...  Driver 15  Driver 16  Driver 17  \
Rider 1         1         1          0  ...          0          1          0   

         Driver 18  Driver 19  Driver 20  Driver 21  Driver 22  Driver 23  \
Rider 1          0          0          1          0          0          0   

         Driver 24  
Rider 1          1  

[1 rows x 24 columns]

In [ ]:
# Decision Variables

In [ ]:
driver = cp.Variable(num_drivers, boolean = True)

In [ ]:
# Objective

In [ ]:
#shortest enroute time 
obj = cp.Minimize(driver@eta.values.T)

In [ ]:
# Constraints 
# MDR does not exceed the driver's 
constraint = [driver@maximum_dispatch_radius.values.T == 1]
#rider is only picked up by 1 driver - but this is automatically set when you set the objective to minimize as there is only 1 smallest value 


In [ ]:
prob = cp.Problem(obj, constraint)

In [ ]:
# Solution in terms of a dispatching recommendation

In [ ]:
prob.solve()

0.05424831544228037

In [ ]:
driver.value

array([0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0.])

**The driver number 8 is matched with the optimal ETA being 0.054**

<h2>II. Batching Dispatch Protocol: Minimize the Aggregated En Route Time Model</h2>
<br>
<font size="+1">
    <ul>
        <li style="color:blue">Recall:</li>
        <br>
        <ul style="color:blue">
            <li>An alternative matching approach is <i>batching</i> where rider requests are collected for a short time window (typically on the order of seconds), at the end of which an optimization problem is solved to pair each request with an open driver.</li>
            <br>
            <li>If there are riders that are not matched in this batch, then they are carried over and used for the optimization problem in the next batching window.</li>
            <br>
            <li>Batching can reduce the rider waiting time relative to the <i>first-dispatch</i> protocol by consolidating requests and making better use of the supply of dispatchable drivers.</li>
            <br>
            <ul>
                <li>Note that the first-dispatch protocol can be viewed as a special case of the batching protocol when each batch consists of exactly one rider request.</li>
                <br>
            </ul>
            <li>Due to its substantial benefits, batching has been implemented widely in major ride-hailing platforms.</li>
            <br>
        </ul>
        <li>Assume we want to implement an improved matching algorithm solution, which is better than the benchmark first-dispatch protocol, that is capable of dispatching multiple drivers, from a collection of available drivers, to a collection of riders.</li>
        <br>
        <ul>
            <li>Note that not every driver will necessarily be dispatched, and not every rider will necessarily be matched with an available driver (it will depend on how the number of riders and number of drivers are related ($\geq, \leq, =$)).</li>
            <br>
        </ul>
        <li>Specifically, we are to implement the so-called <i>batching protocol</i> as an optimization problem given the randomly generated data below.</li>
        <br>
        <li>The data consists of:</li>
        <br>
        <ul>
            <li>A collection of riders that are requesting pick-ups from a collection of drivers.</li>
            <br>
            <li>A data frame consisting of some measure of physical distance from each rider to each driver.</li>
            <br>
            <li>A data frame consisting of the en route time, also known as the expected time of arrival (ETA), which represents the estimated time it will take for a driver to meet a rider once they have been dispatched.</li>
            <br>
        </ul>
      <li><b>Using the given data, formulate an optimization model (decision variables, objective function, constraints) that can be used to implement the <b style="color:blue">batch</b> protocol that seeks to match drivers and riders in a way that <b style="color:blue">minimizes the total en route time (i.e. total ETA).</b></b></li>
</font>

In [ ]:
rng = np.random.default_rng(42)

In [ ]:
# Number of Riders
num_riders = rng.poisson(lam=15)

# Number of drivers
num_drivers = rng.poisson(lam=15)

# Goal is to route DRIVER to RIDER
# Equivalently, to match a rider with a driver

# Physical measure of distance between riders and drivers
distances = pd.DataFrame(rng.uniform(0, 30, (num_riders, num_drivers)),
                         columns=[f'Driver {j}' for j in range(1, num_drivers+1)],
                         index=[f'Rider {i}' for i in range(1,num_riders+1)])

# Temporal measure of distance between riders and drivers
# Expected Time of Arrival
# ETA should be a function of the distance, driver's allowable speed, and some idiosyncratic noise
# (need a more realistic way to assign a variance, maybe as a function of distance)
eta = distances.apply(lambda x: rng.pareto(x)) 

In [ ]:
eta

Driver 1   Driver 2  Driver 3  Driver 4  Driver 5  Driver 6  \
Rider 1   0.063526   0.486780  0.012106  0.027373  0.022479  0.014494   
Rider 2   0.248347   0.877538  0.227651  0.015941  0.065569  0.049331   
Rider 3   0.001538   0.012287  0.251564  0.037826  0.000859  0.020208   
Rider 4   0.067861   0.025888  0.027469  0.067392  0.111707  0.186770   
Rider 5   0.004850   0.107339  0.020358  1.008363  0.031519  1.087487   
Rider 6   1.001942   0.035147  0.024661  0.009173  0.127556  0.079233   
Rider 7   0.003812   0.101547  0.016014  0.065231  0.003376  0.122381   
Rider 8   0.091473  26.841782  0.006282  0.198894  0.013723  0.940601   
Rider 9   0.043976   0.058538  1.595765  0.000632  0.071684  0.004295   
Rider 10  0.682454   0.107257  0.367071  0.019731  0.553485  0.843472   
Rider 11  0.026139   0.089118  0.021173  0.204976  0.060355  0.017654   
Rider 12  0.345731   0.005674  0.092542  0.161446  0.205112  0.027236   
Rider 13  0.120719   0.019797  5.063834  0.184301  0.121341  1.019954   
Rider 14  0.100741   0.018636  0.270171  0.012913  0.004388  0.190199   
Rider 15  0.389356   0.082966  0.262117  0.016489  0.000323  0.265519   
Rider 16  0.028592   0.009005  0.451636  0.064703  0.024193  0.034196   
Rider 17  0.097746   0.054380  0.028725  0.042862  0.042587  0.314049   
Rider 18  0.005569   0.260279  0.101016  0.012954  0.999041  0.010149   

          Driver 7      Driver 8  Driver 9  Driver 10  Driver 11  Driver 12  \
Rider 1   0.118065  6.075330e-01  0.067738   0.944572   0.024951   0.011973   
Rider 2   0.019989  3.700551e-02  0.040100   0.200043   0.174144   0.023295   
Rider 3   0.218212  4.846250e+06  0.041857   0.121579   0.001143   0.018190   
Rider 4   0.672072  2.396151e-02  0.132433   0.015500   0.090009   0.037363   
Rider 5   0.004370  3.867961e-02  0.011159   0.071111   1.429397   0.028954   
Rider 6   0.060852  5.473993e-02  0.020133   0.054029   0.267101   0.013120   
Rider 7   0.051385  5.924328e-03  0.230812   0.049417   0.042624   0.138119   
Rider 8   0.340011  1.959335e-01  0.002736   0.273332   0.173345   0.084653   
Rider 9   0.030670  1.964556e-02  0.152024   0.019887   0.080725   0.273153   
Rider 10  0.027580  8.747272e-03  0.002220   0.053902   0.037098   0.036860   
Rider 11  0.044325  1.135187e-02  0.087258   0.024700   0.420585   0.134623   
Rider 12  0.129432  1.056465e-03  0.074379   0.008900   0.115787   0.521314   
Rider 13  0.391547  2.321312e-03  0.003724   0.001099   0.071222   0.009563   
Rider 14  0.171434  3.029475e-02  0.004416   0.061765   0.087587   0.142730   
Rider 15  0.108395  8.245641e-02  1.467781   0.162683   0.854522   0.299350   
Rider 16  0.218574  8.515662e-02  0.236526   0.017663   0.118924   0.015749   
Rider 17  0.101288  1.600290e-01  0.027763   0.067973   0.001266   0.030170   
Rider 18  0.006512  8.466872e-02  0.064391   0.023520   0.014773   0.049757   

          Driver 13  Driver 14  Driver 15   Driver 16  Driver 17  Driver 18  
Rider 1    0.013173   0.308573   0.028032    0.169770   0.109341   0.560946  
Rider 2    0.310061   0.197105   0.133713    0.032297   0.001956   0.167547  
Rider 3    0.098644   0.079491   0.212200    0.530337   0.092002   0.184825  
Rider 4    0.114589   0.259361   0.183641    0.045809   0.008773   0.007093  
Rider 5    0.045596   0.046097   0.077539    0.039752   0.007066   0.705123  
Rider 6    0.277899   0.024043   0.172007    0.033663   0.238994   0.016490  
Rider 7    0.293317   0.048901   0.148144    0.000879   0.035532   0.339986  
Rider 8    0.153671   0.180374   0.080745  427.269935   0.065623   0.025025  
Rider 9    0.160231   0.014484   0.080289    0.043899   0.410271   0.034298  
Rider 10   0.038825   0.047305   0.208807    0.149172   0.050306   0.121838  
Rider 11   0.004170   0.034830   0.036467    1.551704   8.111048   0.097883  
Rider 12   0.008955   0.049329   0.464615    0.111481   0.000157   0.049817  
Rider 13   0.048446   0.002279   0.232928    0.157595   0.009333   0.045471  
Rider 14   0.086161   0.2040

In [ ]:
distances

Driver 1   Driver 2   Driver 3   Driver 4   Driver 5   Driver 6  \
Rider 1    3.843409  13.511578  11.123941  27.802950  19.315954  24.682848   
Rider 2   14.001630   1.314113   4.628685  20.491469  22.342865  29.025292   
Rider 3   24.967794  24.142931  11.624351   8.649843  20.474865   4.192575   
Rider 4   16.957083  22.949966  19.041550  16.607382  16.776215   9.118503   
Rider 5   19.929406  12.191606  24.420612   5.009188   0.681362   2.701436   
Rider 6    3.540177  28.856930  27.257421  20.991214   7.976099  29.075291   
Rider 7   22.755586  21.583889  12.962791  18.819265  17.522939  19.495398   
Rider 8   17.727465   0.684116  28.756776  14.469103  23.482057   2.481900   
Rider 9    4.207473  16.621084   3.257272  20.167203   8.437014  19.782679   
Rider 10   8.727535  15.451714   7.678953  28.081307   4.938235   1.347319   
Rider 11  22.403688   7.873815  28.104395   7.229117   3.682738  24.933380   
Rider 12   3.954667  20.329760   3.654975  15.189898  20.827873  17.433498   
Rider 13   8.593387  27.744253   0.745785  16.655941  19.019253   3.176922   
Rider 14   6.979222  15.923088  18.180475  26.032169  18.093215  12.377147   
Rider 15   4.385197  24.739926   9.310040   4.316158  27.629114   4.965952   
Rider 16  26.250148  25.533949   1.304252   5.444952   7.102346   7.481627   
Rider 17  29.387120  24.060776  23.384326  19.274498  23.369891   4.036566   
Rider 18  10.991773   9.824867  11.383922  20.572300   8.906294  28.465738   

           Driver 7   Driver 8   Driver 9  Driver 10  Driver 11  Driver 12  \
Rider 1   13.302426   6.817162  16.637544   1.914518  24.828935  18.949932   
Rider 2    9.774761  11.113791  14.086674   5.684141   3.897645  14.271148   
Rider 3    5.997246   0.220868  23.607731  19.945526  21.154961  23.421871   
Rider 4    0.924535  13.101522   6.437540  12.255859  25.602092   7.018185   
Rider 5   21.670781  13.856317   4.838153  15.031343   4.569363  20.889611   
Rider 6   23.362527  21.506706  13.480845   8.167247   2.891729  27.078072   
Rider 7    2.533330  12.474222   1.248425  14.819725   9.895836   4.335726   
Rider 8   14.599750  14.721210  28.134794  17.151842  14.204682   8.009270   
Rider 9   21.809838  23.059425   3.232228  27.480355   6.906420   1.122377   
Rider 10  13.052912  29.771267  26.750318  22.458241  26.723775  26.803399   
Rider 11   4.598529   5.378049  17.981484  26.236861   5.893040   9.309710   
Rider 12   5.993270  24.123736  21.462214  22.169520   3.931733   3.712614   
Rider 13   4.210188  12.573430  28.986957  17.881277  27.990697  24.130827   
Rider 14  11.225521  12.776463  19.557931  26.024719  13.616906   7.435187   
Rider 15   8.541602   4.608402   3.464702   0.634440   1.661862   5.239244   
Rider 16  17.136980  12.487873   1.477624  11.208424  15.712588   3.050157   
Rider 17  16.082041  15.426686  25.727164  13.883981  11.552685  19.186898   
Rider 18  27.490441  14.427313   9.850836  16.063044  25.456815  19.577620   

          Driver 13  Driver 14  Driver 15  Driver 16  Driver 17  Driver 18  
Rider 1   22.742632  10.635779  29.120941  26.793634  23.351505   5.839161  
Rider 2    6.807280  20.094420  13.114558  24.980346  21.007953   9.370999  
Rider 3   13.767473  17.062236   4.193910   3.435902  20.052089  14.132886  
Rider 4    1.749082   8.441517   8.807813  19.857495  16.710965  23.516946  
Rider 5   13.384688  11.430637   9.045363  18.908478  10.854378   2.629498  
Rider 6   13.673289   6.070901   9.178699  17.376587   5.303183  25.698429  
Rider 7    3.102089  17.629337   5.117789  27.753604  17.431834  10.406094  
Rider 8    9.947070  15.620172  13.167344   0.648362  24.788758  26.884823  
Rider 9   16.645574  11.127669  24.893692  24.247544   9.514167  28.586982  
Rider 10  15.565751   9.477872  23.160373  19.849838  11.209732   2.834000  
Rider 11  23.322145  29.154793  15.022236   4.316925   0.418089   6.889681  
Rider 12  27.826877  11.927346   9.028461  14.657521  19.885926  28.668698  
Rider 13  14.021448  23.542903   0.535104   

In [ ]:
# Decision Variables

In [ ]:
driver = cp.Variable((num_riders, num_drivers), boolean = True)
driver

Variable((18, 18), boolean=True)

In [ ]:
eta.iloc[2]

Driver 1     1.538345e-03
Driver 2     1.228711e-02
Driver 3     2.515636e-01
Driver 4     3.782597e-02
Driver 5     8.586307e-04
Driver 6     2.020840e-02
Driver 7     2.182124e-01
Driver 8     4.846250e+06
Driver 9     4.185670e-02
Driver 10    1.215787e-01
Driver 11    1.142687e-03
Driver 12    1.818995e-02
Driver 13    9.864424e-02
Driver 14    7.949099e-02
Driver 15    2.122003e-01
Driver 16    5.303366e-01
Driver 17    9.200186e-02
Driver 18    1.848251e-01
Name: Rider 3, dtype: float64

In [ ]:
# Objective: Minimize eta 
#driver multiply by rider inverse transpose 
total_eta = 0
for i in range(num_riders):
  total_eta += driver[i] @ eta.iloc[[i]].values.T

obj = cp.Minimize(total_eta)

In [ ]:
# Constraints 
# Constraint 1 (take into consideration number of drivers and riders)
if num_drivers <= num_riders:
  constraint1 = [cp.sum(driver) == num_drivers]
else: 
  constraint1 = [cp.sum(driver) == num_riders]
#Constraint 2: 1 driver can only pick up 1 rider
for i in range(num_drivers):
  constraint1.append(sum(driver[:,i]) <=1)
#Constraint 3: 1 rider can only be picked up by 1 driver
for j in range(num_riders):
  constraint1.append(sum(driver[j,:]) <= 1)

In [ ]:
prob = cp.Problem(obj, constraint1)

In [ ]:
# Solution in terms of a dispatching recommendation

In [ ]:
prob.solve()

0.10146319689847205

In [ ]:
pd.DataFrame(driver.value)

0    1    2    3    4    5    6    7    8    9    10   11   12   13   14  \
0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
1   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2   1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
5   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
6   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
7   0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
8   0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
9   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
11  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
12  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0   
13  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
14  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
15  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
17  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   

     15   16   17  
0   0.0  0.0  0.0  
1   0.0  1.0  0.0  
2   0.0  0.0  0.0  
3   0.0  0.0  1.0  
4   0.0  0.0  0.0  
5   0.0  0.0  0.0  
6   1.0  0.0  0.0  
7   0.0  0.0  0.0  
8   0.0  0.0  0.0  
9   0.0  0.0  0.0  
10  0.0  0.0  0.0  
11  0.0  0.0  0.0  
12  0.0  0.0  0.0  
13  0.0  0.0  0.0  
14  0.0  0.0  0.0  
15  0.0  0.0  0.0  
16  0.0  0.0  0.0  
17  0.0  0.0  0.0

**Interpretation: The minimum total ETA for all riders and drivers is 0.10146319689847205. 
The matching is shown in the above table, with the first driver and rider being index of 0**

<h2>III. Batching Dispatch Protocol: Maximize Profit Model with a Maximum Dispatch Radius</h2>
<br>
<font size="+1">
    <ul>
        <li style="color:blue">Recall:</li>
        <br>
        <ul style="color:blue">
            <li>When batch matching is used by ride-hailing platforms, the dispatched driver could miss the opportunity of being matched with a rider that they drive passed while en route to their matched rider.</li>
            <br>
            <li>To mitigate this inefficiency, some ride-hailing platforms proposed a <i>maximum dispatch radius</i> (MDR)  constraint with carefully selected thresholds.</li>
            <br>
            <li>A maximum dispatch radius constraint means that a dispatch is made only if the en route time is below the threshold.</li>
            <br>
            <li>This prevents the driver from being dispatched to pick up very distant riders and experiencing an opportunity cost by not matching with future possible riders that are much closer.</li>
            <br>
        </ul>
        <li>Assume we want to implement a variant on the batching dispatch protocol that minimized total ETA of all matches.</li>
        <br>
        <li><b>This new batching protocol seeks to <b style="color:red">maximize the total profit</b> coming from all matched rider-driver pairs, while ensuring that the matching respects a maximum dispatch radius constraint.</b></li>
        <br>
        <ul>
            <li>Note that not every driver will necessarily be dispatched, and not every rider will necessarily be matched with an available driver (it will depend on how the number of riders and number of drivers are related ($\geq, \leq, =$)).</li>
            <br>
        </ul>
        <li>Specifically, we are to implement a variation on the <i>batching protocol</i> (as described above) as an optimization problem given the randomly generated data below.</li>
        <br>
        <li>The data consists of:</li>
        <br>
        <ul>
            <li>A collection of riders that are requesting pick-ups from a collection of drivers.</li>
            <br>
            <li>A data frame consisting of some measure of physical distance from each rider to each driver.</li>
            <br>
            <li>A maximum dispatch radius (MDR) threshold and a data frame signalling when a rider-driver combination satisfies the MDR constraint.</li>
            <br>
            <li>A data frame consisting of the profits per trip if there is a match between riders and drivers.</li>
            <br>
            <ul>
                <li>Profits can vary based on surge pricing conditions in different locations. That is, surge pricing depends not only on the dynamics of supply and demand, but also on geospatial locations of the rider requests.</li>
                <br>
            </ul>
        </ul>
        <li><b>Using the given data, formulate an optimization model (decision variables, objective function, constraints) that can be used to implement the <b style="color:blue">batch</b> protocol that seeks to match drivers and riders in a way that <b style="color:blue">maximizes the total profits while satisfying the MDR constraints.</b></b></li>
        <br>
      
</font>

In [ ]:
rng = np.random.default_rng(42)

In [ ]:
# Number of Riders
num_riders = rng.poisson(lam=15)

# Number of drivers
num_drivers = rng.poisson(lam=15)

# Goal is to route DRIVER to RIDER
# Equivalently, to match a rider with a driver

# Physical measure of distance between riders and drivers
distances = pd.DataFrame(rng.uniform(0, 30, (num_riders, num_drivers)),
                         columns=[f'Driver {j}' for j in range(1, num_drivers+1)],
                         index=[f'Rider {i}' for i in range(1,num_riders+1)])

# Map distances to a binary constraint which consists 1 if within MDR and 0 otherwise
# A value of 1 means the driver and rider can be connected.
MDR = 15 # don't dispatch outside of MDR miles
maximum_dispatch_radius = (distances < MDR).apply(lambda x: np.where(x, 1, 0)) # np.where(distances < MDR, 1, 0)

# Temporal measure of distance between riders and drivers
# Expected Time of Arrival
# ETA should be a function of the distance, driver's allowable speed, and some idiosyncratic noise
# need a more realistic way to assign a variance, maybe as a function of distance
profits = distances.apply(lambda x: x + rng.lognormal(mean=1, sigma=3)) 

In [ ]:
profits

Driver 1   Driver 2    Driver 3   Driver 4    Driver 5   Driver 6  \
Rider 1   32.262213  15.045836  102.386560  53.660612  659.617014  49.027790   
Rider 2   42.420434   2.848371   95.891304  46.349131  662.643925  53.370233   
Rider 3   53.386598  25.677189  102.886971  34.507505  660.775926  28.537516   
Rider 4   45.375887  24.484224  110.304169  42.465044  657.077275  33.463444   
Rider 5   48.348210  13.725864  115.683231  30.866850  640.982423  27.046377   
Rider 6   31.958981  30.391188  118.520040  46.848876  648.277159  53.420232   
Rider 7   51.174390  23.118147  104.225411  44.676927  657.824000  43.840339   
Rider 8   46.146268   2.218374  120.019396  40.326765  663.783117  26.826841   
Rider 9   32.626276  18.155342   94.519892  46.024865  648.738074  44.127620   
Rider 10  37.146339  16.985972   98.941572  53.938969  645.239295  25.692260   
Rider 11  50.822492   9.408073  119.367014  33.086779  643.983799  49.278321   
Rider 12  32.373470  21.864018   94.917595  41.047560  661.128934  41.778439   
Rider 13  37.012190  29.278511   92.008404  42.513603  659.320314  27.521863   
Rider 14  35.398026  17.457346  109.443094  51.889831  658.394275  36.722088   
Rider 15  32.804001  26.274184  100.572660  30.173820  667.930175  29.310893   
Rider 16  54.668952  27.068207   92.566872  31.302614  647.403407  31.826568   
Rider 17  57.805924  25.595034  114.646946  45.132160  663.670951  28.381507   
Rider 18  39.410577  11.359125  102.646542  46.429963  649.207355  52.810679   

           Driver 7   Driver 8   Driver 9  Driver 10   Driver 11  Driver 12  \
Rider 1   13.326754   9.040791  16.718324   2.488682  277.903786  37.354547   
Rider 2    9.799089  13.337420  14.167455   6.258305  256.972496  32.675762   
Rider 3    6.021574   2.444497  23.688512  20.519690  274.229813  41.826486   
Rider 4    0.948863  15.325151   6.518321  12.830024  278.676943  25.422799   
Rider 5   21.695109  16.079946   4.918934  15.605508  257.644214  39.294226   
Rider 6   23.386855  23.730335  13.561625   8.741411  255.966580  45.482687   
Rider 7    2.557658  14.697851   1.329206  15.393889  262.970688  22.740340   
Rider 8   14.624078  16.944839  28.215574  17.726006  267.279533  26.413885   
Rider 9   21.834167  25.283054   3.313009  28.054520  259.981271  19.526991   
Rider 10  13.077240  31.994896  26.831098  23.032405  279.798626  45.208014   
Rider 11   4.622858   7.601678  18.062264  26.811026  258.967891  27.714325   
Rider 12   6.017598  26.347365  21.542994  22.743685  257.006584  22.117229   
Rider 13   4.234516  14.797058  29.067738  18.455441  281.065548  42.535442   
Rider 14  11.249849  15.000091  19.638711  26.598883  266.691758  25.839802   
Rider 15   8.565931   6.832031   3.545482   1.208605  254.736714  23.643859   
Rider 16  17.161308  14.711502   1.558404  11.782589  268.787440  21.454772   
Rider 17  16.106369  17.650315  25.807945  14.458145  264.627536  37.591513   
Rider 18  27.514769  16.650942   9.931616  16.637208  278.531666  37.982235   

          Driver 13  Driver 14  Driver 15  Driver 16  Driver 17  Driver 18  
Rider 1   23.076573  10.765658  32.120143  26.864307  23.714480  12.770330  
Rider 2    7.141221  20.224299  16.113760  25.051020  21.370928  16.302168  
Rider 3   14.101414  17.192115   7.193112   3.506576  20.415064  21.064055  
Rider 4    2.083023   8.571396  11.807015  19.928169  17.073940  30.448115  
Rider 5   13.718629  11.560516  12.044565  18.979152  11.217353   9.560666  
Rider 6   14.007230   6.200780  12.177901  17.447261   5.666159  32.629597  
Rider 7    3.436030  17.759216   8.116991  27.824277  17.794809  17.337263  
Rider 8   10.281011  15.750051  16.166546   0.719036  25.151733  33.815992  
Rider 9   16.979515  11.257548  27.892894  24.318218   9.877142  35.518150  
Rider 10  15.899692   9.607751  26.159575  19.920512  11.572707   9.765169  
Rider 11  23.656086  29.284672  18.021438   4.387599   0.781064  13.820849  
Rider 12  28.160817  12.057225  12.027663  14.728195  20.248902 

In [ ]:
distances

Driver 1   Driver 2   Driver 3   Driver 4   Driver 5   Driver 6  \
Rider 1    3.843409  13.511578  11.123941  27.802950  19.315954  24.682848   
Rider 2   14.001630   1.314113   4.628685  20.491469  22.342865  29.025292   
Rider 3   24.967794  24.142931  11.624351   8.649843  20.474865   4.192575   
Rider 4   16.957083  22.949966  19.041550  16.607382  16.776215   9.118503   
Rider 5   19.929406  12.191606  24.420612   5.009188   0.681362   2.701436   
Rider 6    3.540177  28.856930  27.257421  20.991214   7.976099  29.075291   
Rider 7   22.755586  21.583889  12.962791  18.819265  17.522939  19.495398   
Rider 8   17.727465   0.684116  28.756776  14.469103  23.482057   2.481900   
Rider 9    4.207473  16.621084   3.257272  20.167203   8.437014  19.782679   
Rider 10   8.727535  15.451714   7.678953  28.081307   4.938235   1.347319   
Rider 11  22.403688   7.873815  28.104395   7.229117   3.682738  24.933380   
Rider 12   3.954667  20.329760   3.654975  15.189898  20.827873  17.433498   
Rider 13   8.593387  27.744253   0.745785  16.655941  19.019253   3.176922   
Rider 14   6.979222  15.923088  18.180475  26.032169  18.093215  12.377147   
Rider 15   4.385197  24.739926   9.310040   4.316158  27.629114   4.965952   
Rider 16  26.250148  25.533949   1.304252   5.444952   7.102346   7.481627   
Rider 17  29.387120  24.060776  23.384326  19.274498  23.369891   4.036566   
Rider 18  10.991773   9.824867  11.383922  20.572300   8.906294  28.465738   

           Driver 7   Driver 8   Driver 9  Driver 10  Driver 11  Driver 12  \
Rider 1   13.302426   6.817162  16.637544   1.914518  24.828935  18.949932   
Rider 2    9.774761  11.113791  14.086674   5.684141   3.897645  14.271148   
Rider 3    5.997246   0.220868  23.607731  19.945526  21.154961  23.421871   
Rider 4    0.924535  13.101522   6.437540  12.255859  25.602092   7.018185   
Rider 5   21.670781  13.856317   4.838153  15.031343   4.569363  20.889611   
Rider 6   23.362527  21.506706  13.480845   8.167247   2.891729  27.078072   
Rider 7    2.533330  12.474222   1.248425  14.819725   9.895836   4.335726   
Rider 8   14.599750  14.721210  28.134794  17.151842  14.204682   8.009270   
Rider 9   21.809838  23.059425   3.232228  27.480355   6.906420   1.122377   
Rider 10  13.052912  29.771267  26.750318  22.458241  26.723775  26.803399   
Rider 11   4.598529   5.378049  17.981484  26.236861   5.893040   9.309710   
Rider 12   5.993270  24.123736  21.462214  22.169520   3.931733   3.712614   
Rider 13   4.210188  12.573430  28.986957  17.881277  27.990697  24.130827   
Rider 14  11.225521  12.776463  19.557931  26.024719  13.616906   7.435187   
Rider 15   8.541602   4.608402   3.464702   0.634440   1.661862   5.239244   
Rider 16  17.136980  12.487873   1.477624  11.208424  15.712588   3.050157   
Rider 17  16.082041  15.426686  25.727164  13.883981  11.552685  19.186898   
Rider 18  27.490441  14.427313   9.850836  16.063044  25.456815  19.577620   

          Driver 13  Driver 14  Driver 15  Driver 16  Driver 17  Driver 18  
Rider 1   22.742632  10.635779  29.120941  26.793634  23.351505   5.839161  
Rider 2    6.807280  20.094420  13.114558  24.980346  21.007953   9.370999  
Rider 3   13.767473  17.062236   4.193910   3.435902  20.052089  14.132886  
Rider 4    1.749082   8.441517   8.807813  19.857495  16.710965  23.516946  
Rider 5   13.384688  11.430637   9.045363  18.908478  10.854378   2.629498  
Rider 6   13.673289   6.070901   9.178699  17.376587   5.303183  25.698429  
Rider 7    3.102089  17.629337   5.117789  27.753604  17.431834  10.406094  
Rider 8    9.947070  15.620172  13.167344   0.648362  24.788758  26.884823  
Rider 9   16.645574  11.127669  24.893692  24.247544   9.514167  28.586982  
Rider 10  15.565751   9.477872  23.160373  19.849838  11.209732   2.834000  
Rider 11  23.322145  29.154793  15.022236   4.316925   0.418089   6.889681  
Rider 12  27.826877  11.927346   9.028461  14.657521  19.885926  28.668698  
Rider 13  14.021448  23.542903   0.535104   

In [ ]:
# maximum_dispatch_radius

In [ ]:
driver = cp.Variable((num_riders, num_drivers), boolean = True)
driver

Variable((18, 18), boolean=True)

In [ ]:
# Objective
#driver multiply by rider inverse transpose 
total_profits = 0
for i in range(num_riders):
  total_profits += driver[i] @ profits.iloc[[i]].values.T
#maximize the profits
obj = cp.Maximize(total_profits)

In [ ]:
# Constraints 
# Constraint 1 (take into consideration number of drivers and riders)
if num_drivers <= num_riders:
  constraint = [cp.sum(driver) == num_drivers]
else: 
  constraint = [cp.sum(driver) == num_riders]
#Constraint 2: 1 driver can only pick up 1 rider
for i in range(num_drivers):
  constraint.append(sum(driver[:,i]) <=1)
#Constraint 3: 1 rider can only be picked up by 1 driver
for j in range(num_riders):
  constraint.append(sum(driver[j,:]) <= 1)

#constraint4: MDR has to be 1 
for i in range(num_riders):
  constraint.append(driver[i] @maximum_dispatch_radius.iloc[[i]].values.T == 1)

In [ ]:
prob = cp.Problem(obj, constraint)

In [ ]:
# Solution in terms of a dispatching recommendation

In [ ]:
prob.solve()

1322.7378169500578

In [ ]:
pd.DataFrame(driver.value)

0    1    2    3    4    5    6    7    8    9    10   11   12   13   14  \
0   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1   1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3   0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
5   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
6   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0   
7   0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
8   0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
9   0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
11  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
12  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
13  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
14  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
15  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
17  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

     15   16   17  
0   0.0  0.0  0.0  
1   0.0  0.0  0.0  
2   0.0  0.0  1.0  
3   0.0  0.0  0.0  
4   0.0  0.0  0.0  
5   0.0  0.0  0.0  
6   0.0  0.0  0.0  
7   0.0  0.0  0.0  
8   0.0  0.0  0.0  
9   0.0  0.0  0.0  
10  0.0  0.0  0.0  
11  1.0  0.0  0.0  
12  0.0  0.0  0.0  
13  0.0  0.0  0.0  
14  0.0  1.0  0.0  
15  0.0  0.0  0.0  
16  0.0  0.0  0.0  
17  0.0  0.0  0.0

**Interpretation: The maximum total profit to aim at is 1322.74. The matching is shonw in the table above, with the index 0 being the first rider and driver**

<h2>IV. Optimal Scheduling of Uber Push Notifications</h2>
<br>
<font size="+1">
    <ul>
        The data consists of:</li>
        <br>
        <ul>
            <li>The time periods (time slots) where a push notification can be sent, as well as the number of push notifications that can be sent.</li>
            <br>
            <li>A data frame (scores) consisting of some measure of the value of a <b>(push, time)</b> scheduling pair.</li>
            <br>
        </ul>
        <li style="color:dodgerblue">A few remarks:</li>
        <br>
        <ul style="color:dodgerblue">
            <li>Given a set of candidate push notifications for a user and a set of possible delivery times, the optimization framework should identify the optimal <b>(push, time)</b> pairs.</li>
            <br>
            <li><b>We will ignore all window and expiriation constraints given in the article.</b></li>
            <br>
            <li>Not all notifications need to be scheduled, only the ones that lead to the highest profit.</li>
            <br>
        </ul>

</font>


**Uber's business motivation for implementing the push notification:**
  * Increase engagement and retention among its users
  * Keep the users informed about promotions and discounts --> They can be reminded of the app --> Use it more 
  * Increase Uber's revenue 

**Where machine learning would be used in the solution to this push notification scheduling problem:**
  * Predict the optimal times to send notifications to users by understanding and analyzing users' behaviors on the application. From there, we can identify the best times to enable notifications that will maximize audience engagement
  * The data we might need: users' past behavior and any patterns of their interactions with the app, user survey and feedback. 
  * Then, we can create the model that learn those patterns that could maximize the outcomes which is the user engagement on the app and send the push notifications at the time that the model predicts. 


In [ ]:
rng = np.random.default_rng(42)

In [ ]:
# Number of times that you can schedule a push notification
time_periods = rng.integers(1,40) # 4

# Number of push notifications you need to schedule
num_notifications = rng.integers(1,40) # 9

# Goal is to schedule NOTIFICATION to TIME PERIOD

# Score that measures the value of a (push, time) pair
# It represents the output of a machine learning model 
# that predicts the probability of a user making an order within 24 hours of receiving push i at time t
scores = pd.DataFrame(rng.uniform(0, 50, (num_notifications, time_periods)),
                         columns=[f'Time Slot {j}' for j in range(1, time_periods+1)],
                         index=[f'Push Notification {i}' for i in range(1,num_notifications+1)])

In [ ]:
scores

Time Slot 1  Time Slot 2  Time Slot 3  Time Slot 4
Push Notification 1     21.943922    42.929896    34.868401     4.708867
Push Notification 2     48.781118    38.056985    39.303215     6.405682
Push Notification 3     22.519297    18.539901    46.338249    32.193256
Push Notification 4     41.138081    22.170710    11.361936    27.729239
Push Notification 5      3.190863    41.381559    31.583220    37.904387
Push Notification 6     17.726298    48.534901    44.656056    38.919175
Push Notification 7      9.731935    23.336050     2.190188     7.714475
Push Notification 8     34.152448    37.238108    48.375487    16.291268
Push Notification 9     18.522985    23.477791     9.473568     6.496075
Push Notification 10    23.785246    11.345467    33.490700    21.857596
Push Notification 11    41.633910    35.013255    15.618332    41.612990
Push Notification 12    40.238218    19.373919    14.416405    34.124775
Push Notification 13     6.987624     9.995410     0.368113    39.346219
Push Notification 14    33.242543    35.258269    39.036452    22.945789
Push Notification 15    28.437060     6.989850     5.726504    33.420148
Push Notification 16    23.554810    28.261805    38.249943    31.735916
Push Notification 17    27.678970    27.960358    15.197505     1.540892
Push Notification 18    21.835869    10.729234    20.426432    42.670154
Push Notification 19    11.696974     2.915137    14.069195    14.679688
Push Notification 20    33.095826    27.851608    39.194910    33.215677
Push Notification 21    20.319343    40.701019     8.348646     1.135604
Push Notification 22     4.502393    36.117968    23.093862     8.063589
Push Notification 23    25.052239     7.615605    34.816019    22.307814
Push Notification 24    19.051061    15.075604    31.514130    18.090631
Push Notification 25     4.382496     5.900295    48.094883    45.429035
Push Notification 26    34.985357    13.293498    48.458819    38.937545
Push Notification 27    35.844509    22.468075    13.612078     4.819548
Push Notification 28    45.130120    22.788814    10.118168    15.297831
Push Notification 29    28.960978     8.838639    42.830714    37.925976
Push Notification 30    35.973148    21.604652    31.365442    29.204898
Push Notification 31    32.492330     4.222216    20.790370     2.080709

In [ ]:
#decision variables
pushnoti = cp.Variable((num_notifications, time_periods), boolean = True)
pushnoti

Variable((31, 4), boolean=True)

In [ ]:
# Objective
#timeslots multiply by pushnoti inverse transpose 
total_scores = 0
for i in range(num_notifications):
   total_scores += pushnoti[i] @ scores.iloc[[i]].values.T
#maximize total scores
obj = cp.Maximize(total_scores)

In [ ]:
# Constraints 
# Constraint 1 (take into consideration number of notifications vs number of timeslots)
if time_periods <= num_notifications:
  constraint1 = [cp.sum(pushnoti) == time_periods]
else: 
  constraint1 = [cp.sum(pushnoti) == num_notifications]
#Constraint 2: 1 push noti can only be shown in 1 time slot
for i in range(time_periods):
  constraint1.append(sum(pushnoti[:,i]) <=1)
#Constraint 3: 1 time slot can only show 1 push noti
for j in range(num_notifications):
  constraint1.append(sum(pushnoti[j,:]) <= 1)

In [ ]:
prob = cp.Problem(obj, constraint1)

In [ ]:
# Solution in terms of a dispatching recommendation

In [ ]:
prob.solve()

191.20387220434952

In [ ]:
pd.DataFrame(pushnoti.value)

0    1    2    3
0   0.0  0.0  0.0  0.0
1   1.0  0.0  0.0  0.0
2   0.0  0.0  0.0  0.0
3   0.0  0.0  0.0  0.0
4   0.0  0.0  0.0  0.0
5   0.0  1.0  0.0  0.0
6   0.0  0.0  0.0  0.0
7   0.0  0.0  0.0  0.0
8   0.0  0.0  0.0  0.0
9   0.0  0.0  0.0  0.0
10  0.0  0.0  0.0  0.0
11  0.0  0.0  0.0  0.0
12  0.0  0.0  0.0  0.0
13  0.0  0.0  0.0  0.0
14  0.0  0.0  0.0  0.0
15  0.0  0.0  0.0  0.0
16  0.0  0.0  0.0  0.0
17  0.0  0.0  0.0  0.0
18  0.0  0.0  0.0  0.0
19  0.0  0.0  0.0  0.0
20  0.0  0.0  0.0  0.0
21  0.0  0.0  0.0  0.0
22  0.0  0.0  0.0  0.0
23  0.0  0.0  0.0  0.0
24  0.0  0.0  0.0  1.0
25  0.0  0.0  1.0  0.0
26  0.0  0.0  0.0  0.0
27  0.0  0.0  0.0  0.0
28  0.0  0.0  0.0  0.0
29  0.0  0.0  0.0  0.0
30  0.0  0.0  0.0  0.0

**The maximum total scores is 191.2. The matchings are: noti 2 and time slot 1, noti 6 at time slot 2, noti 26 at time slot 3, noti 25 at time slot 4**